src/java/org/apache/cassandra/journal/Flusher.java

```diff
@@ -155,21 +158,23 @@ public void doRun(Interruptible.State state) throws InterruptedException
         private void doFlush()
         {
             journal.selectSegmentToFlush(segmentsToFlush);
+            segmentsToFlush.sort(comparing(s -> s.descriptor));
+
             // only schedule onSuccess callbacks for a segment if the preceding segments
             // have been fully flushed, to preserve 1:1 mapping between record's position
             // in the journal and onSuccess callback scheduling order
             boolean scheduleOnSuccessCallbacks = true;
             try
             {
-                for (ActiveSegment<K> segment : segmentsToFlush)
+                for (ActiveSegment<K, V> segment : segmentsToFlush)
                 {
                     try
                     {
                         scheduleOnSuccessCallbacks = doFlush(segment, scheduleOnSuccessCallbacks) && scheduleOnSuccessCallbacks;
```

Comment:
See the comment above. Renamed the method to reflect its real intention - to tell if the segment is now fully completed *and* flushed. It will be flushed again in the future, in the next iteration following the completion of the last write to that segment.